### PR TITLE
Add country mapping overwrite field on geoip maxmind filter

### PR DESCRIPTION
**Pull request by barroca**
Commit Message:

[Add country map file to ovewrite country code.](https://github.com/envoyproxy/envoy/commit/e50606feee33bb3f31c2a7f2ca29252dee081dd0) 

Added :ref:`Geoip country mapping path option<envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when needed. The country mapping file is a json containing the iso code to be mapped from:to so it is possible to override the result read from the Maxmind database. If the path is incorrect or the json is invalid, a warning will be logged, the metric country_mapping.parse_error will be incrementedand an empty map will be used.

Additional Description:
I've added a commit that changes the behaviour to throw exceptions instead of logging errors when the country mapping file is invalid. Feel free to ask for removal if you prefer to have it just sending warnings.

Risk Level:
This PR changes the maxmind geoip filter and should have low risk since it doesn't change other filter behaviours. 

Testing:
Using curl with `curl -k -H "X-Forwarded-For: 85.24.143.2, 40.0.0.1" -H "X-Real-IP: 1.1.1.10"`  might be enough to validate the configuration.

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: changelogs/current.yaml [hunk 1]
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
     Removed ``envoy.reloadable_features.enable_aws_credentials_file`` runtime flag and legacy code paths.
 
 new_features:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
+    needed. The country mapping file is a json containing the iso code to be mapped from:to so it is possible to override the result
+    read from the Maxmind database. If the path is incorrect or the json is invalid, a warning will be logged, the metric
+    country_mapping.parse_error will be incremented and an empty map will be used.
```
Comment (phlax) [inline changelogs/current.yaml:123]:
```suggestion
    Added :ref:`Geoip country mapping path option
    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
    needed. The country mapping file is JSON containing the ISO code to be mapped from:to so it is possible to override the result
    read from the Maxmind database. If the path is incorrect or the JSON is invalid, a warning will be logged, the metric
    ``country_mapping.parse_error`` will be incremented and an empty map will be used.
```

**Review by phlax — COMMENTED** [review]
thanks for working on this @barroca - minor changelog nits ...

File: changelogs/current.yaml [hunk 2]
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
     Removed ``envoy.reloadable_features.enable_aws_credentials_file`` runtime flag and legacy code paths.
 
 new_features:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```
Comment (phlax) [inline changelogs/current.yaml:120]:
```suggestion
    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```

File: api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto [hunk 3]
```diff
@@ -39,4 +40,11 @@ message MaxMindConfig {
   // Common provider configuration that specifies which geolocation headers will be populated with geolocation data.
   common.v3.CommonGeoipProviderConfig common_provider_config = 4
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config country mappings overrides, e.g. /etc/country-mapping.conf.
+  // File expected to have .conf extension. File should be a json with the following format:
+  // {
+  //   "country_code": "country_code" # comment
+  // }
```
Comment (phlax) [inline api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:48]:
```suggestion
  // The file is expected to have a ``.conf`` extension, and should be JSON with the following format:
  //
  // .. code-block::json
  //   {
  //     "country_code": "country_code" # comment
  //   }
  //
```

not 100% if this will resolve the current docs build issues but i think we want to do this anyway

File: api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto [hunk 4]
```diff
@@ -39,4 +40,11 @@ message MaxMindConfig {
   // Common provider configuration that specifies which geolocation headers will be populated with geolocation data.
   common.v3.CommonGeoipProviderConfig common_provider_config = 4
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config country mappings overrides, e.g. /etc/country-mapping.conf.
```
Comment (phlax) [inline api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:44]:
```suggestion
  // Optional full file path to the config country mappings overrides, e.g. ``/etc/country-mapping.conf``.
```

**Review by phlax — COMMENTED** [review]
some fixes/suggestions for docs

File: api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto [hunk 5]
```diff
@@ -39,4 +40,15 @@ message MaxMindConfig {
...
   common.v3.CommonGeoipProviderConfig common_provider_config = 4
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config for country mapping overrides, e.g. ``/etc/country-mapping.conf``.
+  // The file is expected to have a ``.conf`` extension, and should be JSON with the following format:
+  //
+  // .. code-block::json
+  //   {
+  //     "country_code": "country_code" # comment
+  //   }
```
Comment (phlax) [inline api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:50]:
```suggestion
  //   }
  //
```

**Comment by phlax** [comment]
@barroca the docs will render once they are able to - theres still an issue with the changelog link

File: changelogs/current.yaml [hunk 6]
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
     Removed ``envoy.reloadable_features.enable_aws_credentials_file`` runtime flag and legacy code paths.
 
 new_features:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```
Comment (phlax) [inline changelogs/current.yaml:120]:
```suggestion
    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path>` to allow override of country code when
```

File: api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto [hunk 7]
```diff
@@ -39,4 +40,16 @@ message MaxMindConfig {
   // Common provider configuration that specifies which geolocation headers will be populated with geolocation data.
   common.v3.CommonGeoipProviderConfig common_provider_config = 4
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config for country mapping overrides, e.g. ``/etc/country-mapping.conf``.
+  // The file is expected to have a ``.conf`` extension, and should be JSON with the following format:
+  //
+  // .. code-block::json
+  //   {
+  //     "country_code": "country_code" # comment
```
Comment (ravenblackx) [inline api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:49]: Probably worth clarifying that it's not actually json because json doesn't support comments.

I might also suggest using `//` comments rather than `#` comments, adding an optional ignore_comments argument in the json loader (to be used [here](https://github.com/envoyproxy/envoy/blob/main/source/common/json/json_internal.cc#L728) as the argument to [sax_parse](https://json.nlohmann.me/api/basic_json/sax_parse/)) and supporting comments that way ([nlohmann's blurb about comments](https://json.nlohmann.me/features/comments/))

That would be better tested than the regex which would, for example, have a surprising outcome if someone were to try to use some sort of meta country codes like
```
{
  "#X": "boop"
}
```
(The hacky regex method as currently coded will transform this into invalid json.)

Comment (phlax) [reply api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:49]: seems the json is not rendering - im guessing the comment is why (https://storage.googleapis.com/envoy-pr/bdfa297/docs/api-v3/extensions/geoip_providers/maxmind/v3/maxmind.proto.html)

if the comment is necessary you can add `:force:` to the json code-block, otherwise maybe just remove the comment

also wondering if the json code-block needs a space - ie `.. code-block:: json`

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 8]
```diff
@@ -111,6 +126,62 @@ MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_
...
+    std::string content_without_comments = compiled_regex->replaceAll(file_data, "");
+    // Load as json and fill country_mapping_.
+    Json::ObjectSharedPtr json = Json::Factory::loadFromString(content_without_comments);
+    json->iterate([&](const std::string& key, const Json::Object& value) {
+      country_mapping->insert(std::make_pair(key, value.asString()));
+      return true;
+    });
+  }
+  END_TRY
+  catch (Json::Exception&) {
```
Comment (ravenblackx) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:168]: I think `END_TRY CATCH` might be the expected thing here, in support of a build with disabled exceptions. ([example](https://sourcegraph.com/github.com/envoyproxy/envoy@4994e59af25898d56f496fece3ed24dacbdfa87b/-/blob/source/common/router/header_parser_utils.cc?L54))

Also, including the message from the exception in the rethrown exception would probably be helpful to the user, rather than just telling them the entire file is wrong.

File: api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto (hunk 7 above)
Comment (barroca) [reply api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto]: made some changes on the parser

Comment (phlax) [reply api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto]: JSON doesnt support any comments - so this is still failing

as said above - if comments really _should_ be included we can force it - but its indicating that (most) json parsers are not going to work with comments included

File: test/extensions/geoip_providers/maxmind/test_data/country_mapping.conf [hunk 9]
```diff
@@ -0,0 +1,6 @@
+// -*- coding: utf-8 -*-
+{
+  "PF": "AU", // French Polynesia
+  "PT-30": "MA", // Madeira
+  "SE": "XX" // Ovewrite for testing country mapping
```
Comment (ravenblackx) [inline test/extensions/geoip_providers/maxmind/test_data/country_mapping.conf:5]: Overwrite

**Review by ravenblackx — DISMISSED** [review]
This looks okay to me except the one super-nit comment typo. (And some bad test style things in the json test but they're consistent with what's already there so I'm not gonna ask you to fix them.)

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 10]
```diff
@@ -27,6 +37,10 @@ GeoipProviderConfig::GeoipProviderConfig(
                                                  : absl::nullopt),
       anon_db_path_(!config.anon_db_path().empty() ? absl::make_optional(config.anon_db_path())
                                                    : absl::nullopt),
+      // Path is hardcoded since we don't have the field on the OSS version of the proto.
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:40]: nit: remove non relevant comment

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 11]
```diff
@@ -111,6 +126,55 @@ MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_
...
+      auto file_or_error = file_system_.fileReadToEnd(country_mapping_path.value());
+
+      if (!file_or_error.status().ok()) {
+        throwEnvoyExceptionOrPanic(fmt::format("Unable to open country mapping file {}. Error {}",
+                                               country_mapping_path.value(),
+                                               std::string(file_or_error.status().message())));
+      }
+      parseCountryMapping(file_or_error.value(), country_mapping_path.value(), country_mapping);
+    } else {
+      config_->incCountryMappingFileParseError();
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:143]: practically, this is not file parse error, but rather file does not exist/could not be resolved by filesystem object.

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 12]
```diff
@@ -150,6 +214,12 @@ void GeoipProvider::lookupInCityDb(
           populateGeoLookupResult(mmdb_lookup_result, lookup_result,
                                   config_->countryHeader().value(), MMDB_COUNTRY_LOOKUP_ARGS[0],
                                   MMDB_COUNTRY_LOOKUP_ARGS[1]);
+          auto it = lookup_result.find(config_->countryHeader().value());
+          std::string country;
+          if (it != lookup_result.end()) {
+            country = it->second;
+          }
+          lookup_result[config_->countryHeader().value()] = getCountryMapping(country);
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:222]: highly likely this assignment operator will copy the string value of `getCountryMapping(country)` call, you could use move semantics (capture getCountryMapping(country) by rvalue) instead to avoid extra copy: https://cplusplus.com/reference/string/string/operator=/

File: test/extensions/geoip_providers/maxmind/config_test.cc [hunk 13]
```diff
@@ -353,6 +402,34 @@ TEST(MaxmindProviderConfigTest, DifferentProviderInstancesForDifferentProtoConfi
...
 
+using MaxmindProviderConfigDeathTest = MaxmindProviderConfigTest;
+
+TEST_F(MaxmindProviderConfigDeathTest, CountryMappingPathDoesNotExist) {
+  const auto provider_config_yaml = R"EOF(
+    common_provider_config:
+      geo_headers_to_add:
+        city: "x-geo-city"
+    city_db_path: %s
+    country_mapping_path: "/usr/share/geoip2Gcs/wrong_path.conf"
```
Comment (nezdolik) [inline test/extensions/geoip_providers/maxmind/config_test.cc:413]: since this is optional config param, we should have a test where this field is not configured.

File: test/extensions/geoip_providers/maxmind/config_test.cc [hunk 14]
```diff
@@ -331,19 +370,29 @@ TEST(MaxmindProviderConfigTest, DifferentProviderInstancesForDifferentProtoConfi
...
-      absl::StrFormat(provider_config_yaml2, city_db_path, anon_db_path);
+      absl::StrFormat(provider_config_yaml2, city_db_path, anon_db_path, country_mapping_path);
   TestUtility::loadFromYaml(processed_provider_config_yaml1, provider_config1);
   TestUtility::loadFromYaml(processed_provider_config_yaml2, provider_config2);
   NiceMock<Server::Configuration::MockFactoryContext> context;
+
+  auto api = Api::createApiForTest();
+
+  Stats::IsolatedStoreImpl stats_store_;
+  Stats::ScopeSharedPtr scope_{stats_store_.createScope("")};
```
Comment (nezdolik) [inline test/extensions/geoip_providers/maxmind/config_test.cc:393]: these variables are local to a single test code, the names should not have underscores in the end.

File: test/extensions/geoip_providers/maxmind/geoip_provider_test.cc [hunk 15]
```diff
@@ -98,6 +108,34 @@ TEST_F(GeoipProviderTest, ValidConfigCityAndIspDbsSuccessfulLookup) {
   expectStats("isp_db");
 }
 
+TEST_F(GeoipProviderTest, ValidConfigCityAndIspDbsSuccessfulLookupWithoutResultDontCauseCrash) {
```
Comment (nezdolik) [inline test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:111]: could you elaborate more/add explanatory comment how this test relates to newly added country mapping feature? E.g. why could there be a crash

File: test/extensions/geoip_providers/maxmind/geoip_provider_test.cc [hunk 16]
```diff
@@ -265,6 +304,74 @@ TEST_F(GeoipProviderTest, ValidConfigCityMultipleLookups) {
   expectStats("city_db", 2, 2);
 }
 
+TEST_F(GeoipProviderTest, ValidCountryMapping) {
```
Comment (nezdolik) [inline test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:307]: nit: rename test to smth like `ValidCountryMappingCountryValueOverriden` etc

Comment (nezdolik) [reply test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:307]: also consider splitting 2 cases in this test into 2 tests

File: test/extensions/geoip_providers/maxmind/geoip_provider_test.cc [hunk 17]
```diff
@@ -287,6 +394,40 @@ TEST_F(GeoipProviderDeathTest, GeoDbNotSetForConfiguredHeader) {
                "performing lookups");
 }
 
+TEST_F(GeoipProviderDeathTest, CountryMappingFileStatusNotOk) {
```
Comment (nezdolik) [inline test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:397]: i may have missed it, but is there is a test when file path does not exist?

**Review by nezdolik — COMMENTED** [review]
Thank you for the contribution! Please update integration test as well

**Comment by ravenblackx** [comment]
Now that nezdolik has the comments I'm going to assign it over. :)

**Comment by nezdolik** [comment]
Apologies for delayed review, i have been swamped past weeks.

**Comment by barroca** [comment]
@nezdolik I addressed your comments, but right now I'm having trouble identifying the leak that I'm getting from the presubmit  job https://dev.azure.com/cncf/envoy/_build/results?buildId=163920&view=logs&j=1439b9f7-a348-5b50-b5fe-ea612ea91241&t=1002ac43-da84-5fae-70b2-98833b702d09

I'm still investigating, but if you have any clue let me know.

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 18]
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
   ENVOY_LOG(debug, "Shutting down Maxmind geolocation provider");
   if (city_db_) {
     MMDB_close(city_db_.get());
+    city_db_ = nullptr;
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:106]: why is this change needed? Database pointer is a unique smart pointer owned by provider instance, memory will be released by smart pointer on geoip provider destruction.

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 19]
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
...
   }
   if (isp_db_) {
     MMDB_close(isp_db_.get());
+    isp_db_ = nullptr;
   }
   if (anon_db_) {
     MMDB_close(anon_db_.get());
+    anon_db_ = nullptr;
+  }
+  if (country_mapping_) {
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:116]: why this is `if` block is needed?

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc [hunk 20]
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
...
+    country_mapping_->clear();
+    country_mapping_ = nullptr;
   }
 }
 
 MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_path) {
   if (db_path) {
-    MMDB_s maxmind_db;
-    int result_code = MMDB_open(db_path.value().c_str(), MMDB_MODE_MMAP, &maxmind_db);
+    MMDB_s* maxmind_db = new MMDB_s;
```
Comment (nezdolik) [inline source/extensions/geoip_providers/maxmind/geoip_provider.cc:124]: please revert to previous version, Envoy uses smart pointers for memory management, this is the reason you got a memory leak.

File: source/extensions/geoip_providers/maxmind/geoip_provider.cc (hunk 19 above)
Comment (barroca) [reply source/extensions/geoip_providers/maxmind/geoip_provider.cc:116]: country_mapping_ is a pointer as anon_db_ and others. keeping the same standard

Comment (nezdolik) [reply source/extensions/geoip_providers/maxmind/geoip_provider.cc:113]: country_mapping_ is smart pointer to C++ `absl::map`, there is no need to clean it within the destructor of the owning object (provider). While all db types are smart pointers to C struct types, in C there are no destructors and maxmind library recommends to call `MMDB_close` to release dynamically allocated memory.
